src/bms/command.rs

```diff
@@ -294,3 +328,151 @@ impl TryFrom<u8> for LnMode {
         })
     }
 }
+
+/// Associates between object `K` and [`ObjId`] with memoization.
+/// It is useful to assign object ids for many objects with its equality.
+pub struct ObjIdManager<'a, K: ?Sized> {
+    value_to_id: HashMap<&'a K, ObjId>,
+    used_ids: HashSet<ObjId>,
+    unused_ids: VecDeque<ObjId>,
+}
+
+impl<'a, K: ?Sized> Default for ObjIdManager<'a, K>
+where
+    K: std::hash::Hash + Eq,
+{
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<'a, K: ?Sized> ObjIdManager<'a, K>
+where
+    K: std::hash::Hash + Eq,
+{
+    /// Create a new empty ObjIdManager
+    #[must_use]
+    pub fn new() -> Self {
+        let unused_ids: VecDeque<ObjId> = ObjId::all_values().collect();
+
+        Self {
+            value_to_id: HashMap::new(),
+            used_ids: HashSet::new(),
+            unused_ids,
+        }
+    }
+
+    /// Create a new ObjIdManager with iterator of assigned entries
+    pub fn from_entries<I: IntoIterator<Item = (&'a K, ObjId)>>(iter: I) -> Self {
+        let mut value_to_id: HashMap<&'a K, ObjId> = HashMap::new();
+        let mut used_ids: HashSet<ObjId> = HashSet::new();
+
+        // Collect all entries first
+        let entries: Vec<_> = iter.into_iter().collect();
+
+        // Mark used IDs and build the mapping
+        for (key, assigned_id) in entries {
+            value_to_id.insert(key, assigned_id);
+            used_ids.insert(assigned_id);
+        }
+
+        let unused_ids: VecDeque<ObjId> = ObjId::all_values()
+            .filter(|id| !used_ids.contains(id))
+            .collect();
+
+        Self {
+            value_to_id,
+            used_ids,
+            unused_ids,
+        }
+    }
+
+    /// Returns whether the key is already assigned any id
+    pub fn is_assigned(&self, key: &'a K) -> bool {
+        self.value_to_id.contains_key(key)
+    }
+
+    /// Get or allocate an ObjId for a key without creating tokens
+    pub fn get_or_new_id(&mut self, key: &'a K) -> ObjId {
+        if let Some(&id) = self.value_to_id.get(key) {
+            id
+        } else if let Some(new_id) = self.unused_ids.pop_front() {
+            self.used_ids.insert(new_id);
+            self.value_to_id.insert(key, new_id);
+            new_id
+        } else {
+            ObjId::null()
+        }
+    }
```

Comment:
I'd like `get_or_new_id` to return the `Option<ObjId>` to handle it by the callee when failed the allocate.
```suggestion
    pub fn get_or_new_id(&mut self, key: &'a K) -> Option<ObjId> {
        if let Some(&id) = self.value_to_id.get(key) {
            Some(id)
        } else if let Some(new_id) = self.unused_ids.pop_front() {
            self.used_ids.insert(new_id);
            self.value_to_id.insert(key, new_id);
            Some(new_id)
        } else {
            None
        }
    }
```